### ISSUE TITLE
web.xmlrpc returns response with  "text/html" content-type when xmlrpc method does not exist.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2430]
|<img alt="Remi's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| Remi reported|
|-|-|
|Trac ID|trac#2430|
|Type|defect|
|Created|2007-02-02 11:09:57Z|

When you call an xmlrpc method that doesn't excist, the content type of the response is "text/html" and not "text/xml". The problem lies in the xmlrpc.XMLRPC render method. 

When an exception is raised on self.getFunction(functionPath),  self._cbRender(f, request) is called while the content-type is the default "text/html".

Solution is simple, move the "request.setHeader("content-type", "text/xml")" line above the try.

- Remi Cool -
```python
def render(self, request):
        request.content.seek(0, 0)
        args, functionPath = xmlrpclib.loads(request.content.read())
        try:
            function = self._getFunction(functionPath)
        except Fault, f           
            self._cbRender(f, request)
        else:
            request.setHeader("content-type", "text/xml")
            defer.maybeDeferred(function, *args).addErrback(
                self._ebRender
            ).addCallback(
                self._cbRender, request
            )
        return server.NOT_DONE_YET
```

<details><summary>Searchable metadata</summary>

```
trac-id__2430 2430
type__defect defect
reporter__Remi Remi
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web web
keywords__xmlrpc xmlrpc
time__1170414597000000 1170414597000000
changetime__1174155343000000 1174155343000000
version__None None
owner__ 
cc__therve cc__exarkun cc__dreid cc__oubiwann
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve removed owner|
|-|-|

Ready to review in xmlrpc-content-type-2430.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_3"></a>|@dreid set owner to @therve|
|-|-|

```
            for factory in self.factories:
                if 'Content-type' in factory.headers:
                    self.assertEquals(factory.headers['Content-type'],
                                      'text/xml')
```

It seems like this if statement is unecessary, and maybe even dangerous.  If it were ever possible that at this point the content-type were not set, this test would still pass.  That is highly unlikely but the if statement isn't necessary for the test to pass so why risk it?

Otherwise this is really good.  I'd prefer it if these tests didn't hit the network, but I won't hold that against you.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve set owner to @dreid|
|-|-|

Replying to [dreid](#note_3):
> It seems like this if statement is unecessary, and maybe even dangerous.  If it were ever possible that at this point the content-type were not set, this test would still pass.  That is highly unlikely but the if statement isn't necessary for the test to pass so why risk it?

True, that should be ok now.

> Otherwise this is really good.  I'd prefer it if these tests didn't hit the network, but I won't hold that against you.

Yeah, these tests may need a big refactoring...

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @therve|
|-|-|

```test_errors``` has a grammatical error:  _errors path_ should be _error paths_.  I'd expand on this a bit, though.  How about:

```
Verify that for each way a method exposed via XML-RPC can fail, the
correct 'Content-Type' header is set in the response and that the
client-side Deferred is errbacked with an appropriate C{Fault}
instance.
```

A minor point.  I think the newlines in the XMLRPCTestAuthenticated test methods which were reformatted would be somewhat better off like this:

```
    def test_authInfoInURL(self):
        p = xmlrpc.Proxy("http://%s:%s@...:%d/" % (self.user,
            self.password, self.port))
        return p.callRemote("authinfo").addCallback(
            self.assertEquals, [self.user, self.password])
```

But actually the appropriate fix for this is probably something more along the lines of:

```
    def test_authInfoInURL(self):
        url = "http://%s:%s@...:%d/" % (
            self.user, self.password, self.port)
        p = xmlrpc.Proxy(url)
        d = p.callRemote("authinfo")
        d.addCallback(self.assertEquals, [self.user, self.password])
        return d
```

Which I think most people would find more readable than any of the other versions.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve removed owner|
|-|-|

Thanks, I did the corrections.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set owner to @therve|
|-|-|

It looks like this branch was made when trunk was in a bad state.  Tests are failing on Win32.  Could you merge this branch forward?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve removed owner|
|-|-|

Merged forward in xmlrpc-content-type-2430-2.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @glyph<br>@glyph set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_13"></a>|@glyph set owner to @therve<br>@glyph set status to `new`|
|-|-|

I can see the previous review feedback was dealt with, and I think the change looks good.

The only thing that slightly bothers me about it is the use of the case-sensitive ```'Content-type'``` string.  Not only is it weird to be using a case-sensitive comparison API there, its case disagrees with the docstring.

I don't think that this branch is the right place to fix twisted.web's wonky header capitalization, and this test being extra strict _does_ at least alert us to further case-corruption issues in the header, so please just make the docstring's case agree for now, and merge.

(I would guess that buggy things that care about header-case generally want to see the RFC-style "Content-Type" case rather than "Content-type".  I could swear Twisted had some changes made to it at one point to deal with that issue, but I can't find them now.)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_14"></a>|@glyph commented|
|-|-|

(Improving header markup for posterity.)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_17"></a>|@therve set status to `closed`|
|-|-|

(In [19828]) Merge xmlrpc-content-type-2430-2

Author: therve
Reviewers: dreid, exarkun, glyph
Fixes [#2430](https://github.com/twisted/twisted/issues/2430)

Fix twisted.web.xmlrpc content-type of responses to always be 'text/xml', and
add tests to this behavior.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_18"></a>|@therve commented|
|-|-|

(In [19829]) Revert [19828]

Failures in win32-py2.5 slave. Refs [#2430](https://github.com/twisted/twisted/issues/2430).

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_19"></a>|@therve set status to `reopened`|
|-|-|

Strange failures:

```
Traceback (most recent call last):
  File "e:\twistedbuildbot\W32-full2.5-scmikes-select\Twisted\twisted\web\test\test_xmlrpc.py", line 189, in cb
    self.assertEquals(factory.headers['Content-type'],
exceptions.KeyError: 'Content-type'
===============================================================================
[ERROR]: twisted.web.test.test_xmlrpc.XMLRPCTestAuthenticated.test_errors

Traceback (most recent call last):
Failure: twisted.web.test.test_xmlrpc.TestValueError: 
===============================================================================
[ERROR]: twisted.web.test.test_xmlrpc.XMLRPCTestAuthenticated.test_errors

Traceback (most recent call last):
  File "e:\twistedbuildbot\W32-full2.5-scmikes-select\Twisted\twisted\internet\defer.py", line 107, in maybeDeferred
    result = f(*args, **kw)
  File "e:\twistedbuildbot\W32-full2.5-scmikes-select\Twisted\twisted\web\test\test_xmlrpc.py", line 67, in xmlrpc_fail
    raise TestRuntimeError
twisted.web.test.test_xmlrpc.TestRuntimeError: 
===============================================================================
[ERROR]: twisted.web.test.test_xmlrpc.XMLRPCTestCase.test_errors

Traceback (most recent call last):
  File "e:\twistedbuildbot\W32-full2.5-scmikes-select\Twisted\twisted\web\test\test_xmlrpc.py", line 189, in cb
    self.assertEquals(factory.headers['Content-type'],
exceptions.KeyError: 'Content-type'
===============================================================================
[ERROR]: twisted.web.test.test_xmlrpc.XMLRPCTestCase.test_errors

Traceback (most recent call last):
  File "e:\twistedbuildbot\W32-full2.5-scmikes-select\Twisted\twisted\internet\defer.py", line 107, in maybeDeferred
    result = f(*args, **kw)
  File "e:\twistedbuildbot\W32-full2.5-scmikes-select\Twisted\twisted\web\test\test_xmlrpc.py", line 67, in xmlrpc_fail
    raise TestRuntimeError
twisted.web.test.test_xmlrpc.TestRuntimeError: 
===============================================================================
[ERROR]: twisted.web.test.test_xmlrpc.XMLRPCTestCase.test_errors

Traceback (most recent call last):
Failure: twisted.web.test.test_xmlrpc.TestValueError: 
===============================================================================
[ERROR]: twisted.web.test.test_xmlrpc.XMLRPCTestCase2.test_errors

Traceback (most recent call last):
  File "e:\twistedbuildbot\W32-full2.5-scmikes-select\Twisted\twisted\web\test\test_xmlrpc.py", line 189, in cb
    self.assertEquals(factory.headers['Content-type'],
exceptions.KeyError: 'Content-type'
===============================================================================
[ERROR]: twisted.web.test.test_xmlrpc.XMLRPCTestCase2.test_errors

Traceback (most recent call last):
  File "e:\twistedbuildbot\W32-full2.5-scmikes-select\Twisted\twisted\internet\defer.py", line 107, in maybeDeferred
    result = f(*args, **kw)
  File "e:\twistedbuildbot\W32-full2.5-scmikes-select\Twisted\twisted\web\test\test_xmlrpc.py", line 67, in xmlrpc_fail
    raise TestRuntimeError
twisted.web.test.test_xmlrpc.TestRuntimeError: 
===============================================================================
[ERROR]: twisted.web.test.test_xmlrpc.XMLRPCTestCase2.test_errors

Traceback (most recent call last):
Failure: twisted.web.test.test_xmlrpc.TestValueError: 
===============================================================================
[ERROR]: twisted.web.test.test_xmlrpc.XMLRPCTestIntrospection.test_errors

Traceback (most recent call last):
  File "e:\twistedbuildbot\W32-full2.5-scmikes-select\Twisted\twisted\web\test\test_xmlrpc.py", line 189, in cb
    self.assertEquals(factory.headers['Content-type'],
exceptions.KeyError: 'Content-type'
===============================================================================
[ERROR]: twisted.web.test.test_xmlrpc.XMLRPCTestIntrospection.test_errors

Traceback (most recent call last):
  File "e:\twistedbuildbot\W32-full2.5-scmikes-select\Twisted\twisted\internet\defer.py", line 107, in maybeDeferred
    result = f(*args, **kw)
  File "e:\twistedbuildbot\W32-full2.5-scmikes-select\Twisted\twisted\web\test\test_xmlrpc.py", line 67, in xmlrpc_fail
    raise TestRuntimeError
twisted.web.test.test_xmlrpc.TestRuntimeError: 
===============================================================================
[ERROR]: twisted.web.test.test_xmlrpc.XMLRPCTestIntrospection.test_errors

Traceback (most recent call last):
Failure: twisted.web.test.test_xmlrpc.TestValueError: 
```

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_20"></a>|@therve set status to `new`|
|-|-|

So, I don't manage to reproduce this. It seems it can comes from the case sensitive problem (like, if I replace Content-type with Content-Type, I get the failures.

I'll wait for the slave to go up to continue investigation.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_21"></a>|@therve removed owner|
|-|-|

I don't know how it's possible, but 'Content-type' was actually spelled 'Content-Type' on this slave. This is completely insane (bug in capitalize() ? Some other strange thing ?).

I've workaround the problem using a .lower() call.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_22"></a>|@glyph set owner to @therve|
|-|-|

It occurs to me that capitalization is locale-dependent :).  I have no idea how bogdog managed to get a locale that screwed up on his windows machine, but c'est la vie.

Reviewing the branch again I don't see any other issues.

I think it's fine to re-merge.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_23"></a>|@therve set status to `closed`|
|-|-|

(In [19844]) Merge xmlrpc-content-type-2430-2

Author: therve
Reviewers: dreid, exarkun, glyph
Fixes [#2430](https://github.com/twisted/twisted/issues/2430)

Fix twisted.web.xmlrpc content-type of responses to always be 'text/xml', and
add tests to this behavior.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
